[sen5x] Extend with SEN6X support

mikelawrence opened the pull request:
# What does this implement/fix?

Add support for SEN6X sensors to sen5x component. This includes SEN62, SEN63C, SEN65, SEN66, SEN68 and SEN69C sensors.

- Add required ```type``` configuration variable supporting SEN50, SEN54, SEN55, SEN62, SEN63C, SEN65, SEN66, SEN68 and SEN69C. This is a breaking change for previous installations.
- Add CO₂ sensor. Only on SEN63C, SEN66 and SEN69C models.
  - Add ```auto_self_calibration``` boolean configuration variable. Enabled automation CO₂ calibration.
  - Add ```altitude_compensation``` float configuration variable. When configured altitude compensation is enabled at the specified height.
  - Add ```ambient_pressure_calibration_source``` configuration variable. When configured pressure compensation is enabled and pressure will automatically be read from this sensor.
  - Add ```sen5x.set_ambient_pressure_compensation``` action for CO₂ sensor. Setting the current pressure in hPa will enable pressure compensation. You will need to periodically update the ambient pressure.
  - Add ```sen5x.perform_forced_co2_recalibration``` action for CO₂ sensor. When ```auto_self_calibration``` is false, this action is used to force the current CO₂ level to a known value.
- Add HCHO (Formaldehyde) sensor. Only on SEN68 and SEN69C models.
- Add ```sen5x.activate_heater``` action for Humidity sensor. New for SEN62, SEN63C, SEN65, SEN66, SEN68 and SEN69C sensors. This is similar to to the SHT4x heater without the automatic capability.

Notes:
* Going from 3 supported models to 9 requires more model logic.
* SEN6X models do not support Auto Fan Cleaning.
* SEN6X Temperature compensation is different from SEN5x. The follow-on PR (see above) adds temperature compensation and acceleration.

This PR is based on PR esphome/esphome#8318 by @AckPeng. His original work supported the SEN66 only. I added support for all SEN6X models. His preference was to make a new PR instead of integrating my changes.

This PR has been tested on the following models: SEN50, SEN54, SEN55, SEN63C, SEN65 and SEN66. These are the currently available models.

Will add these missing features in a later PR

- Temperature Compensation Action necessary extend slot support in the SEN6X sensors.
- Temperature Acceleration. New feature in SEN6X sensors


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes esphome/esphome/issues/13441
- fixes esphome/feature-requests/issues/2995

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5041

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: sen5x
    id: sen54
    type: sen66
    pm_1_0:
      name: "PM <1µm Weight concentration"
      id: pm_1_0
      accuracy_decimals: 1
    pm_2_5:
      name: "PM <2.5µm Weight concentration"
      id: pm_2_5
      accuracy_decimals: 1
    pm_4_0:
      name: "PM <4µm Weight concentration"
      id: pm_4_0
      accuracy_decimals: 1
    pm_10_0:
      name: "PM <10µm Weight concentration"
      id: pm_10_0
      accuracy_decimals: 1
    temperature:
      name: "Temperature"
      accuracy_decimals: 1
    humidity:
      name: "Humidity"
      accuracy_decimals: 0
    voc:
      name: "VOC"
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    temperature_compensation:
      offset: 0
      normalized_offset_slope: 0
      time_constant: 0
    nox:
      name: "NOₓ"
      accuracy_decimals: 0
    co2:
      name: "CO2"
      accuracy_decimals: 0
      auto_self_calibration: false
    update_interval: 10s

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mikelawrence committed:
Add CONF_ALTITUDE_COMPENSATION to const.py

mikelawrence committed:
Add SEN6X support to sen5x component

mikelawrence committed:
clang-format cleanup

mikelawrence commented:
> Creating a new sen6x component isn't better than just adding it to the sen5x? 

The original PR esphome/esphome#8318 started by integrating into the sen5x component. These components are very similar, especially to the end user. The original sen5x component had eight sensors all of which are available in the SEN6X series. The SEN6X series adds two new sensors (CO2 and HCHO).

The scd4x component supports the SCD40, SCD41 and SCD43 but those components are really only different in performance not interface. Same goes for sht4x and sgp4x components.

I have no problem converting this into a separate sen6x component if that's the consensus.

mikelawrence committed:
Fix CO2 Altitude Compensation

Mistakenly set the CO2 Automatic Self Calibration value.

mikelawrence committed:
Switch to CONF_AUTOMATIC_SELF_CALIBRATIONS

This is already available in consts.py

mikelawrence committed:
Fix CO2 Automatic Self Calibration not reporting correctly in dump_config()

jesserockz commented on esphome/components/scd30/sensor.py:
```diff
@@ -35,8 +36,6 @@
     "ForceRecalibrationWithReference", automation.Action
 )
 
-CONF_ALTITUDE_COMPENSATION = "altitude_compensation"
```
> Create a separate PR if/when you wish to move a constant into `esphome/const.py`.

https://developers.esphome.io/contributing/code/#esphome-specific-idiosyncrasies

mikelawrence replied:
Removed from this PR

jesserockz commented on esphome/components/sen5x/automation.h:
```diff
@@ -17,5 +42,15 @@ template<typename... Ts> class StartFanAction : public Action<Ts...> {
+
+ protected:
+  SEN5XComponent *sen5x_;
+};
```
Not sure why you didn't follow the pattern and use Parented here as well?

mikelawrence replied:
Converted all actions to Parented

jesserockz commented on esphome/components/sen5x/sen5x.h:
```diff
@@ -98,36 +101,58 @@ class SEN5XComponent : public PollingComponent, public sensirion_common::Sensiri
     temp_comp.time_constant = time_constant;
     temperature_compensation_ = temp_comp;
   }
+  void set_co2_auto_calibrate(bool value) { co2_auto_calibrate_ = value; }
```
> All uses of class members and member functions should be prefixed with this-> to distinguish them from global functions/variables.

https://developers.esphome.io/contributing/code/#c-code-style

mikelawrence replied:
Should I make this change on all parts of the code? 

Basically this is the same question as above. I was afraid of making changes out of the scope of adding support for the SEN6X sensors.

kbx81 replied:
If there are missing `this->` elsewhere, feel free to add them as it's a trivial change that doesn't impact other components/files.

mikelawrence replied:
The sen5x.h was riddled with missing ```this->```. Fixed all of them.

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -4,149 +4,184 @@
+void SEN5XComponent::internal_setup_(uint8_t state) {
+  uint16_t string_number[16] = {0};
+  switch (state) {
+    case 0:
```
state really should be an enum type here please.

mikelawrence replied:
Fixed as suggested

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -424,23 +754,225 @@ bool SEN5XComponent::write_temperature_compensation_(const TemperatureCompensati
+  }
+}
+
+Sen5xType SEN5XComponent::str_to_model_(const char *product_name) {
```
This function can be static inline and doesnt need to be part of the class.

Also, if the original value is already a std::string, you might as well just use `const std::string &product_name` and do direct `==` compares in this function.

mikelawrence replied:
Fixed as suggested

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -331,79 +524,216 @@ void SEN5XComponent::update() {
+  return result;
+}
+
+std::string SEN5XComponent::convert_to_string_(const uint16_t array[], uint8_t length) {
```
This can also be static inline and doesn't need to be on the class. Generally if there is no access to `this->` and all work is done with the args then it doesn't need to be on the class.

mikelawrence replied:
Fixed as suggested

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -331,79 +524,216 @@ void SEN5XComponent::update() {
+}
+
+std::string SEN5XComponent::convert_to_string_(const uint16_t array[], uint8_t length) {
+  std::string new_string;
+  const uint16_t *current_int = array;
+  char current_char;
+  // 2 ASCII bytes are encoded in an int
+  do {
+    // first char
+    current_char = *current_int >> 8;
+    if (current_char) {
+      new_string.push_back(current_char);
+      // second char
+      current_char = *current_int & 0xFF;
+      if (current_char) {
+        new_string.push_back(current_char);
+      }
+    }
+    current_int++;
+  } while (current_char && --length);
+  return new_string;
```
Could you not just use `reinterpret_cast<char *>(array)` here to convert the incoming data pointer then create the string?

mikelawrence replied:
Let me look at this a bit closer. It may be wrong.

mikelawrence replied:
```reinterpret_cast<char *>(array)``` worked but I did have to add a loop to convert the array of uint16_t's  to big endian the reinterpret_cast. This is tested and I believe it should work across multiple platforms.

``` c++

static inline std::string convert_to_string_(uint16_t array[], uint8_t length) {
  for (int i = 0; i < length; i++) {
    array[i] = convert_big_endian(array[i]);
  }
  std::string new_string = reinterpret_cast<const char *>(array);
  return new_string;
}

```

jesserockz commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -331,79 +524,216 @@ void SEN5XComponent::update() {
-  if (!this->write_command(SEN5X_CMD_READ_MEASUREMENT)) {
+  if (!this->write_command(cmd)) {
     this->status_set_warning();
-    ESP_LOGD(TAG, "write error read measurement (%d)", this->last_error_);
+    ESP_LOGD(TAG, "Failed to write Measurement Command (last_error=%d)", this->last_error_);
     return;
   }
-  this->set_timeout(20, [this]() {
-    uint16_t measurements[8];
 
-    if (!this->read_data(measurements, 8)) {
+  this->set_timeout(20, [this, length]() {
+    uint16_t measurements[9];
+
+    if (!this->read_data(measurements, length)) {
       this->status_set_warning();
       ESP_LOGD(TAG, "read data error (%d)", this->last_error_);
       return;
     }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_1_0 = 0x%.4x", measurements[0]);
     float pm_1_0 = measurements[0] == UINT16_MAX ? NAN : measurements[0] / 10.0f;
+    if (this->pm_1_0_sensor_ != nullptr) {
+      this->pm_1_0_sensor_->publish_state(pm_1_0);
+    }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_2_5 = 0x%.4x", measurements[1]);
     float pm_2_5 = measurements[1] == UINT16_MAX ? NAN : measurements[1] / 10.0f;
+    if (this->pm_2_5_sensor_ != nullptr) {
+      this->pm_2_5_sensor_->publish_state(pm_2_5);
+    }
 
+    // supported by all
     ESP_LOGVV(TAG, "pm_4_0 = 0x%.4x", measurements[2]);
     float pm_4_0 = measurements[2] == UINT16_MAX ? NAN : measurements[2] / 10.0f;
+    if (this->pm_4_0_sensor_ != nullptr) {
+      this->pm_4_0_sensor_->publish_state(pm_4_0);
+    }
```
There is no point doing maths that wont be used
```suggestion
    if (this->pm_4_0_sensor_ != nullptr) {
      float pm_4_0 = measurements[2] == UINT16_MAX ? NAN : measurements[2] / 10.0f;
      this->pm_4_0_sensor_->publish_state(pm_4_0);
    }
```

mikelawrence replied:
Reworked all of these so that logging and maths are only executed if the sensor exists.

jesserockz commented on esphome/components/sen5x/sensor.py:
```diff
@@ -99,6 +142,7 @@ def float_previously_pct(value):
     cv.Schema(
         {
             cv.GenerateID(): cv.declare_id(SEN5XComponent),
+            cv.Optional(CONF_MODEL): cv.enum(SEN5X_MODELS, upper=True),
```
Should this become a required field to make things easier?
We can then validate the users can only set the sensors the model provides, the actions (by way of different empty Class that simply extend the base class) can also then be validated against the correct model only.

mikelawrence replied:
That was my first thought but decided against it because it would be a breaking change. If you think otherwise I'll definitely change it to required and do configuration validation in the sensor.py.

I've done some looking but couldn't find an example of validating actions, do you know of a component that does this?

mikelawrence replied:
I like making ```model``` required so I made the changes. Configuration Sensor and variables are now validated at compile time. All runtime checks have been removed.

I do not understand what you are asking for when it comes to actions. I have left these as they were with validations at runtime.

jesserockz replied:
Thanks, We can update the actions in another PR to not hold up this one as that is kinda a bit more of an involved change.

jesserockz reviewed (changes requested):
Please also look over https://developers.esphome.io/architecture/logging/

mikelawrence committed:
Merge branch 'esphome:dev' into dev

mikelawrence committed:
Revert "Add CONF_ALTITUDE_COMPENSATION to const.py"

This reverts commit 0f55a19a5f2e0dea182a78674f95022e4cbbe283.

mikelawrence committed:
Merge branch 'dev' of https://github.com/mikelawrence/esphome into dev

mikelawrence committed:
Convert setup state machine state to enumeration

Fix missing set_timeout() in SM_SET_ACI state

mikelawrence committed:
Convert all actions to Parented

Consolidate SEN5X_VALUE_ACTION_SCHEMA schema for actions with a value

mikelawrence committed:
Convert functions to static inline and not on the class

convert_to_string_ changed to reinterpret_cast<const char *>(array)
str_to_model_ refactored using strings
model_to_str_ no changes

mikelawrence committed:
Logging and math is done only when sensor exists

mikelawrence committed:
class members prefixed with this->

mikelawrence committed:
Logging cleanup

mikelawrence committed:
More Logging cleanup

Reverted some of the previous logging
Additional dump_config() optimization

mikelawrence committed:
model is now required in configuration

Add sensor and variable validation at compile time.
Minor Logging change.
convert_to_string_() changed to convert_big_endian() from byteswap to support architectures other than ESP32
Fix syntax errors

mikelawrence committed:
pylint suggestion, convert to mode in {set}

mikelawrence committed:
Add model to test config

mikelawrence committed:
Minor logging change

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'esphome:dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Fix read measurement error on SEN5X models

Recent testing of the SEN50, SEN54 and SEN55 models showed an error in read measurements. Along with fixing this issue, now the data read from a read measurement command matches the number of sensors in the model reducing i2c reads

Added a new verbose log entry when a read measurement fails

sort esphome.const imports

mikelawrence committed:
add @mikelawrence as codeowner

mikelawrence committed:
Fix ruff formatting error

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Fix clang-tidy errors

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mikelawrence committed:
Fix another clang-tidy error

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
CONF_ALTITUDE_COMPENSATION i snow in const.py

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
ruff changes; combine `if` statements using `and`

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Use ESP_LOG_MSG_COMM_FAIL for logging and minor logging cleanup

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

washley commented:
FWIW, I just used this on a sen66 and it worked great. Thanks for the PR.

mikelawrence commented:
@washley Thanks for testing the PR! Hopefully they will merge it soon.

I've running three SEN66 sensors around my house and plan to add more.

mikelawrence committed:
Merge branch 'dev' into dev

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
Merged changes from PR #9894. Tested successfully!

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

Scarpetta3 commented:
Hello there,
I have test this PR but I am experiencing a bad behavior when, after few hours to some days of work, the sensor start to send NaN values for al the metrics (PM, VOC, CO2, ecc.)

any advice for that? In the logs I can’t see anything out of the ordinary. Simply values that start to go NaN

thank you very much in advance.
Regards
Matteo

washley commented:
> I could also write an automation that restarts it every time I receive NaNs N times in a row, but I'd rather not do that because I'm convinced it would shorten the life of the sensor and the ESP32

Not that having a watchdog workaround is a good solution, but why do you believe an occasional reboot would shorten the life of the sensor or the esp32?

washley commented:
> I could also write an automation that restarts it every time I receive NaNs N times in a row, but I'd rather not do that because I'm convinced it would shorten the life of the sensor and the ESP32

Not that having a watchdog workaround is a good solution, but why do you believe an occasional reboot would shorten the life of the sensor or the esp32?

mikelawrence committed:
Merge branch 'dev' into dev

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mikelawrence committed:
Fix merge error, remove duplicate  serial_number_

mikelawrence committed:
Merge branch 'dev' into dev

kbx81 commented on esphome/components/sen5x/sen5x.h:
```diff
@@ -92,46 +112,65 @@ class SEN5XComponent : public PollingComponent, public sensirion_common::Sensiri
+  uint8_t firmware_minor_{0xFF};
   bool store_baseline_;
+  uint32_t seconds_since_last_store_;
+  uint16_t co2_ambient_pressure_{0};
```
Please keep all variables ordered by widest to narrowest types (as they were prior to these changes). Larger types/classes should be last (hence the prior placement of `string` at the bottom).
This approach minimizes padding, reducing memory use and binary size.

mikelawrence replied:
Thanks for the help! These changes should be what you are looking for. 

I did test these changes.

mikelawrence replied:
I'm not sure of the etiquette here, do I "re-request review" just let you come back to this in your own time?

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Order variables appropriately

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

danielkent-net commented:
Just wanted to chime in that it looks like this works on RP2040 platforms as well. My YAML config for testing with an RPi Pico W connected to a SEN66 via I2C0 on GPIO 16 and 17:

```
substitutions:
  devicename: testjig02
  device_name: test_jig_02
  human_device_name: "Test Jig 2"

esphome:
  name: ${devicename}
  comment: ${human_device_name}

external_components:
    - source: github://pr#9254
      components: [ sen5x ]

mdns:
  disabled: false

logger:
  level: DEBUG
  baud_rate: 115200

rp2040:
  board: rpipicow

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  reboot_timeout: 0s
  ap:
    ssid: "${devicename}_fallback"
    password: !secret ap_password
    ap_timeout: 120min


api:
  encryption:
    key: !secret encryption_key
  id: ${devicename}

ota:
  platform: esphome
  password: !secret ota_key


i2c:
  sda: GPIO16
    scl: GPIO17
  scan: false
  id: bus_i2c0

sensor:
  - platform: sen5x
    id: sen66_01
    model: sen66
    pm_1_0:
      name: "PM <1µm Weight concentration"
      id: pm_1_0
      accuracy_decimals: 1
    pm_2_5:
      name: "PM <2.5µm Weight concentration"
      id: pm_2_5
      accuracy_decimals: 1
    pm_4_0:
      name: "PM <4µm Weight concentration"
      id: pm_4_0
      accuracy_decimals: 1
    pm_10_0:
      name: "PM <10µm Weight concentration"
      id: pm_10_0
      accuracy_decimals: 1
    temperature:
      name: "Temperature"
      accuracy_decimals: 1
    humidity:
      name: "Humidity"
      accuracy_decimals: 0
    voc:
      name: "VOC"
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    co2:
      name: "CO2"
      accuracy_decimals: 0
      automatic_self_calibration: false
    address: 0x6B
    update_interval: 10s
```

I'd be happy to help with any further testing if needed!

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
@Scarpetta3 

Glad you were able to figure things out!

mikelawrence commented:
> > One last question: regarding the use of the fan to clean the sensor (I remember that the SEN5X model automatically cleaned the sensor with the fan at maximum speed every week, if I'm not mistaken), how can this be implemented with the SEN66?
> 
> The datasheet for the SEN66 does not indicate that there is an auto cleaning function. There's a service in the `sen5x` component called `sen5x.start_fan_autoclean` that you could call periodically using an esphome [`interval` component](https://esphome.io/components/interval/). It looks like the current code disables measurements during the cleaning cycle, which is also recommended by the datasheet.

I have incorporated your suggestion into the documentation. Thanks! 
And yes the code does follow the recommendations from the datasheet regarding proper use of the heater.  
 
> I can try testing this later today to make sure it works, though I think we should make sure we note this in the documentation.

I tested the yaml from your documentation PR and it works just fine.

``` yaml
 interval:
  - interval: 7d
    then:
      - sen5x.start_fan_autoclean: sen66_01
```

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Update per #11704

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
Hello @sealsix,

There is one big reason not to make this change is the devs grandfathered the max line change count for this PR and I'm worried that even though they did this the PR still may be more than they have time to deal with. The only way I would make changes now is by their request or to fix an issue.

But for discussion sake and maybe a future feature request, you are correct, the sensor will not enter idle mode while operating normally, no matter how you set the update interval. As I understand it the manufacturer intends this sensor to run 24/7 and has a stated life expectancy of 6 - 10 years (as indicated by @danielkent-net ). Is there something in the datasheet suggesting how to implement your mode of operation?

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Back out change from #12219

Allow this PR to continue to work with non-dev versions of ESPHome. I'll add this change back in once 2025.12.0 is released.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
@madpipeline I changed this PR's example yaml to include the NOx sensor. 

For your edification the documentation for this PR is in this [ESPHome Docs  PR](https://github.com/esphome/esphome-docs/pull/5041). Here is a link to documentation preview in that PR [netlify/esphome/deploy-preview](https://deploy-preview-5041--esphome.netlify.app/),  click it and search for "sen5x".

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
@madpipeline I finally had a chance to look at your code. One thing I did notice is you are setting the VOC algorithm tuning parameters to default. Which is fine but I haven't done much with this part of the code and maybe there is a problem in there. These parameters are volatile so just removing the algorithm_tuning section from your ESPHome config is good enough for a test. I didn't see you setting these parameters on the SEN55 so nothing to do there. Remember to wait an hour for VOC readings to stabilize.

sealsix commented:
hello i tested it with SEN66 and think i found bug.
when tried to add both version and uptime text sensors, esphome raises compilation errors. but with added only one of version or uptime text sensor, it works. when i delete (comment) sen5x sensor (for SEN66) setup, it also works. so i think it is related to this PR - sen5x sensor with SEN66 setup.


this sen5x setup yaml is not compilable:
```
esphome:
  name: sen66
  friendly_name: SEN66

esp32:
  board: esp32dev
  framework:
    type: arduino

external_components:
  - source: github://pr#9254
    components: [ sen5x ]

logger:
  hardware_uart: UART0
  level: DEBUG

api:
  encryption:
    key: !secret encryption_key

ota:
  - platform: esphome
    password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

number:
  - platform: template
    id: co2_forced_cal_value
    name: "CO₂ Calibration Value"
    device_class: carbon_dioxide
    entity_category: CONFIG
    optimistic: true
    max_value: 1200
    min_value: 400
    step: 1
    initial_value: 420
    set_action:
      - delay: 1s

i2c:
  id: i2c_sen66
  scan: true
  sda: GPIO21
  scl: GPIO22

interval:
  - interval: 7d
    then:
      - sen5x.start_fan_autoclean: sen66_sensor

# both of them are not compilable
text_sensor:
  - platform: uptime
    name: Uptime
    id: uptime_text_sensor
    format:
      separator: " "

  - platform: version
    name: "ESPHome Verze"
    id: esphome_version_text_sensor

sensor:
  - platform: wifi_signal
    name: "WiFi Signál"
    update_interval: 60s

  - platform: sen5x
    id: sen66_sensor
    model: sen66
    i2c_id: i2c_sen66
    address: 0x6B
    update_interval: 10s
    pm_1_0:
      name: "PM 1"
      id: pm_1_0
      accuracy_decimals: 1
    pm_2_5:
      name: "PM 2.5"
      id: pm_2_5
      accuracy_decimals: 1
    pm_4_0:
      name: "PM 4"
      id: pm_4_0
      accuracy_decimals: 1
    pm_10_0:
      name: "PM 10"
      id: pm_10_0
      accuracy_decimals: 1
    temperature:
      name: "Teplota"
      id: temperature
      accuracy_decimals: 1
    humidity:
      name: "Vlhkost"
      id: humidity
      accuracy_decimals: 0
    voc:
      name: "VOC"
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    co2:
      name: "CO₂"
      automatic_self_calibration: false
      accuracy_decimals: 0
    nox:
      name: "NOx"
# pouze SEN68
#    hcho:
#        name: "Formaldehyd"
```

and raises
```
 - toolchain-xtensa-esp-elf @ 14.2.0+20241119
Reading CMake configuration...
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ off, Compatibility ~ strict
Found 44 compatible libraries
Scanning dependencies...
Dependency Graph
|-- Networking @ 3.3.2
|-- ESPmDNS @ 3.3.2
|-- Update @ 3.3.2
|-- noise-c @ 0.1.10
|-- Wire @ 3.3.2
Building in release mode
Compiling .pioenvs\sen66\src\esphome\components\api\api_connection.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\api_frame_helper.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\api_frame_helper_noise.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\api_pb2.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\api_pb2_service.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\api_server.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\list_entities.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\proto.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\api\subscribe_state.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\esp32\core.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\esp32\gpio.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\esp32\helpers.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\esp32\preferences.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\esphome\ota\ota_esphome.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\i2c\i2c.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\i2c\i2c_bus_arduino.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\logger\logger.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\logger\logger_esp32.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\logger\task_log_buffer.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\md5\md5.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\mdns\mdns_component.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\mdns\mdns_esp32.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\network\util.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\number\automation.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\number\number.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\number\number_call.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\number\number_traits.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\ota\ota_backend.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\ota\ota_backend_arduino_esp32.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\safe_mode\safe_mode.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sen5x\sen5x.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sensirion_common\i2c_sensirion.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sensor\automation.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sensor\filter.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sensor\sensor.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\sha256\sha256.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\socket\bsd_sockets_impl.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\socket\socket.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\template\number\template_number.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\text_sensor\filter.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\text_sensor\text_sensor.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\uptime\text_sensor\uptime_text_sensor.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\wifi\wifi_component.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\wifi\wifi_component_esp_idf.cpp.o
Compiling .pioenvs\sen66\src\esphome\components\wifi_signal\wifi_signal_sensor.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\application.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\color.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\component.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\component_iterator.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\controller_registry.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\entity_base.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\helpers.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\log.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\ring_buffer.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\scheduler.cpp.o
Compiling .pioenvs\sen66\src\esphome\core\util.cpp.o
Compiling .pioenvs\sen66\src\main.cpp.o
Linking .pioenvs\sen66\firmware.elf
C:/Users/seals/.platformio/packages/toolchain-xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/14.2.0/../../../../xtensa-esp-elf/bin/ld.exe: .pioenvs/sen66/src/main.cpp.o:(.literal._Z5setupv+0xf4): undefined reference to `vtable for esphome::version::VersionTextSensor'
C:/Users/seals/.platformio/packages/toolchain-xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/14.2.0/../../../../xtensa-esp-elf/bin/ld.exe: .pioenvs/sen66/src/main.cpp.o:(.literal._Z5setupv+0x234): undefined reference to `esphome::version::VersionTextSensor::set_hide_timestamp(bool)'
C:/Users/seals/.platformio/packages/toolchain-xtensa-esp-elf/bin/../lib/gcc/xtensa-esp-elf/14.2.0/../../../../xtensa-esp-elf/bin/ld.exe: .pioenvs/sen66/src/main.cpp.o: in function `setup()':
C:\Projekty\ostatni\moje\esphome\.esphome\build\sen66/src/main.cpp:298:(.text._Z5setupv+0x638): undefined reference to `esphome::version::VersionTextSensor::set_hide_timestamp(bool)'
collect2.exe: error: ld returned 1 exit status
*** [.pioenvs\sen66\firmware.elf] Error 1
=========================================== [FAILED] Took 57.72 seconds ===========================================
```




without sen5x setup, this yaml is compilable:

```
esphome:
  name: sen66
  friendly_name: SEN66

esp32:
  board: esp32dev
  framework:
    type: arduino


logger:
  hardware_uart: UART0
  level: DEBUG

api:
  encryption:
    key: !secret encryption_key

ota:
  - platform: esphome
    password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password


i2c:
  id: i2c_sen66
  scan: true
  sda: GPIO21
  scl: GPIO22



text_sensor:
  - platform: uptime
    name: Uptime
    id: uptime_text_sensor
    format:
      separator: " "

  - platform: wifi_info
    ip_address:
      name: "${friendly_name} IP Address"

  - platform: version
    name: "ESPHome Verze"
    id: esphome_version_text_sensor
```



and setup with only one version text sensor is compilable (also with only uptime is ok):

```
esphome:
  name: sen66
  friendly_name: SEN66

esp32:
  board: esp32dev
  framework:
    type: arduino

external_components:
  - source: github://pr#9254
    components: [ sen5x ]

logger:
  hardware_uart: UART0
  level: DEBUG

api:
  encryption:
    key: !secret encryption_key

ota:
  - platform: esphome
    password: !secret ota_password

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

number:
  - platform: template
    id: co2_forced_cal_value
    name: "CO₂ Calibration Value"
    device_class: carbon_dioxide
    entity_category: CONFIG
    optimistic: true
    max_value: 1200
    min_value: 400
    step: 1
    initial_value: 420
    set_action:
      - delay: 1s

i2c:
  id: i2c_sen66
  scan: true
  sda: GPIO21
  scl: GPIO22

interval:
  - interval: 7d
    then:
      - sen5x.start_fan_autoclean: sen66_sensor

text_sensor:
  - platform: uptime
    name: Uptime
    id: uptime_text_sensor
    format:
      separator: " "

#  - platform: version
#    name: "ESPHome Verze"
#    id: esphome_version_text_sensor

sensor:
  - platform: wifi_signal
    name: "WiFi Signál"
    update_interval: 60s

  - platform: sen5x
    id: sen66_sensor
    model: sen66
    i2c_id: i2c_sen66
    address: 0x6B
    update_interval: 10s
    pm_1_0:
      name: "PM 1"
      id: pm_1_0
      accuracy_decimals: 1
    pm_2_5:
      name: "PM 2.5"
      id: pm_2_5
      accuracy_decimals: 1
    pm_4_0:
      name: "PM 4"
      id: pm_4_0
      accuracy_decimals: 1
    pm_10_0:
      name: "PM 10"
      id: pm_10_0
      accuracy_decimals: 1
    temperature:
      name: "Teplota"
      id: temperature
      accuracy_decimals: 1
    humidity:
      name: "Vlhkost"
      id: humidity
      accuracy_decimals: 0
    voc:
      name: "VOC"
      algorithm_tuning:
        index_offset: 100
        learning_time_offset_hours: 12
        learning_time_gain_hours: 12
        gating_max_duration_minutes: 180
        std_initial: 50
        gain_factor: 230
    co2:
      name: "CO₂"
      automatic_self_calibration: false
      accuracy_decimals: 0
    nox:
      name: "NOx"
# pouze SEN68
#    hcho:
#        name: "Formaldehyd"
```

mikelawrence commented:
@washley I really like your table, do you mind if I add to it?

| Model   | PM  |  T  | RH  | VOC | NOx | CO2 | HCHO | Supported |   Tested   |
|:---------|:-----:|:---:|:----:|:------:|:-----:|:-----:|:--------:|:------------:|:----------:|
| SEN62   |    ✓  |  ✓ |  ✓  |          |          |         |             |        ✗[^1] |    ✗[^2]    |
| SEN63C |    ✓  |  ✓ |  ✓  |          |         |   ✓   |             |        ✓        |    ✓           |
| SEN65   |    ✓  |  ✓ |  ✓  |    ✓    |   ✓   |        |             |        ✓        |    ✗[^3]    |
| SEN66   |    ✓  |  ✓ |  ✓  |    ✓    |   ✓   |   ✓   |            |        ✓        |    ✓           |
| SEN68   |    ✓  |  ✓ |  ✓  |    ✓    |   ✓   |         |     ✓    |        ✓        |    ✗[^2]    |
| SEN69C |    ✓  |  ✓ |  ✓  |    ✓    |   ✓   |   ✓   |     ✓    |        ✗[^1] |    ✗[^2]    |

[^1]: Datasheet not released.
[^2]: Announced, not yet available.
[^3]: Not working, CRC errors in the Sesirion library.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Refactor logging in dump_config

One less log statement.
Stop dumping info when component is marked failed.

mikelawrence committed:
Fix CO2 not converting unknown value into nan

mikelawrence committed:
Add I2C max_frequency validation

From the datasheet these sensors have a i2c frequency of 100 kHz. This change will make sure your config is 100 kHz or less.

mikelawrence committed:
Enforce 100ms power up i2c communication restriction

sealsix commented:
hello @danielkent-net now it works - i solved it by deleting whole .esphome directory and run build again, sorry i had no idea to try to do this before adding comment :(

mikelawrence commented:
I have a custom [Multi-Sensor](https://github.com/mikelawrence/esphome-indoor-multi-sensor-hardware) which supports both a SGP41 and SEN66. And while normally they aren't used together it can be done. They are 4cm apart. I ran both overnight and this what I ended up with.

|  Sensor |  VOC  |
|:---------|:-------:|
| SEN66  |   208   |
| SGP41  |   149   |

They have an update rate of 1 sec and a 60 sample average and as you can see there is a big difference. There was a similar difference between a SEN55 and SEN66 above. Is it sensor variation? Is it differences in sensor hardware? Is VOC consistent across all Sensirion devices? I think @danielkent-net has it right that it probably not a issue with the underlying code.

mikelawrence commented:
The datasheet has been updated to include the SEN62 and SEN69C. The SEN60 has been dropped and will simplify this component.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
Switching to draft while updating the component to support new models SEN62 and SEN69C while removing SEN60 support.

mikelawrence committed:
Add support for SEN62 and SEN69C. Remove support for SEN60 (no longer in datasheet)

mikelawrence committed:
Merge branch 'dev' into dev

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mikelawrence committed:
Updates to support PR#12567

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Fix buffer size too small in update. Would cause problems with unavailable and untested SEN69C

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Recent change was due to SEN6X datasheet indicating CO2 is an int16. This is likely wrong.

mikelawrence commented:
The last commit reversed a change to the CO2 sensor read measurement. The SEN6X datasheet says the CO2 value is an int16 and if a value of 0X7FFF is returned the CO2 value is actually unknown. This is likely wrong. The c library from Sensirion treats CO2 as a uint16 and the CO2 sensor range is 0 - 40000 which cannot be represented by by an int16. I am also experiencing occasional large spikes in CO2 data that are clearly not real. I've tested this change on both the SEN63C and SEN66. Both report 0xFFFF when first starting up. SEN69C is not available yet. Note there are two different sensors: the SEN66 has a range of 0-40000 while the SEN63C and SEN69C have a range of 0-32000.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Correct unknown values for temperature, humidity and co2

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Fix logic error

mikelawrence committed:
Fix another error

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

madpipeline commented:
>Anybody following this PR know how to force unknown values on temperature or humidity measurements?

I never got unknown values, but did get weird short spikes when I placed the sensor on the kitchen counter, and the sun shined on it ocasionally (SEN66).

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

sealsix commented:
> > Creating a new sen6x component isn't better than just adding it to the sen5x?
> 
> The original PR #8318 started by integrating into the sen5x component. These components are very similar, especially to the end user. The original sen5x component had eight sensors all of which are available in the SEN6X series. The SEN6X series adds two new sensors (CO2 and HCHO).
> 
> The scd4x component supports the SCD40, SCD41 and SCD43 but those components are really only different in performance not interface. Same goes for sht4x and sgp4x components.
> 
> I have no problem converting this into a separate sen6x component if that's the consensus.

i vote also for converting to sen6x component. maybe both components can be refactored to prepare some "base" component and sen5x and 6x will inherit the base code.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Improve Config Validation, temp comp for all, add action blocking logic

mikelawrence committed:
Handle forced CO2 calibrations better. Log errors better and report FRC when finished.

mikelawrence committed:
Merge pull request #3 from mikelawrence/add-sen5x-cv-tc

Merge code improvements

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Mostly cleanup with some code improvements

Follow coding conventions in automation classes.
Include component name in State Machine enum member names.
More consistency with naming config variables and action members.
Use static_cast to make conversion intent clearer in multiple places.
Improve config validation.
Remove unnecessary model logic due to config validation.
Improve logging.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Remove unused ambient_pressure_compensation, fix CO2 to_code

mikelawrence committed:
Fix broken Store Baseline

mikelawrence commented:
Spent time on Store Baselines and discovered the original component is broken is at least three ways.

- Store Baseline uses seconds_since_last_store_ to determine when enough time has elapsed to store the baseline again but it is never updated. Saving to flash would never occur.
- In the update process the code allows up to two additional i2c commands in between writing the get VOC algorithm state and it's associated read data. Not sure if works or not because this section of code is silent on errors but I doubt it.
- It assumes the baseline is two 32-bit number and looks for a difference between the current state and last stored start to determine if it is time it save the baseline. This is how the SGP30, SGP40 and SGP41 work. The issue is Sensirion doesn't define the contents of the 8 byte blob. Note the name changed to algorithm state from baseline.

Recent commits fix Store Baselines.

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Fix clang-tidy

mikelawrence committed:
Remove format_hex_pretty, not all that helpful anyways

mikelawrence committed:
Remove unnecessary truthy returns from set methods

mikelawrence committed:
Refactor ambient pressure compensation

mikelawrence committed:
Rename perform_forced_co2_calibration action to perform_forced_co2_recalibration, this matches the datasheet name for this command

mikelawrence committed:
Revert start_fan_cleaning action to start_fan_autoclean, this was an unnecessary breaking change

mikelawrence committed:
Improve setup error logging

mikelawrence committed:
Simplify model logic in set methods and eliminate unnecessary condition test in setup

mikelawrence committed:
Reorder code to reduce number of line changes. No functionality changes.

mikelawrence committed:
Store Baseline cannot have a default of true as indicated by original sen5x component documentation.

mikelawrence committed:
Convert config validation to typed_schema.
Eliminates model logic in config validation.
Generally easier to read.

mikelawrence committed:
Fix tests, missing type

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Allow 0x7FFF and 0xFFFF as invalid temperature and humidity values for SEN5X sensors. Datasheet seems inconsistent here.

mikelawrence committed:
Ignore learning_time_gain_hours in NOx algorithm_tuning. Per datasheet.

mikelawrence committed:
Merge branch 'dev' of https://github.com/mikelawrence/esphome into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Store Baseline changes
Preference hash based on serial number only
Actions not available when initializing
Ambient Pressure Change needs a greater than 1 hPa change before updating
Logging updates for name consistency

schdro commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -43,369 +86,466 @@ static const LogString *rht_accel_mode_to_string(RhtAccelerationMode mode) {
+  for (int i = 0; i < length; i++) {
+    array[i] = convert_big_endian(array[i]);
+  }
+  std::string new_string = reinterpret_cast<const char *>(array);
```
My first thought was: Could risk some out of bounds read overflow if `array` not NUL terminated. However:
- Datasheets of both SEN5x+SEN6x guarantee NUL termination.
- Also CRC checksums exist (transfer errors = garbage parsing...).

So likely low risk -- more if bad/malicious HW.

Longer run might be topic of replacing with fixed `char[]` (along current memory saving strategy of questioning all `std::string`). Then such `convert_to_string` could also ensure NUL termination.
Which would go more into direction of state before review 2025-07-01 which requested explicitly to replace loop with `reinterpret_cast` ctor.

But IMHO goes too far in such PR, more so when `std::string` existed before.

Only as observation, no req to change.

mikelawrence replied:
I have looked at this multiple times and have considered changing this to `char []` but each time it ends up with more lines added and this as already a big PR. So for now I'm going to keep it the same.

I'm also working on fixing the original sen5x component. There it's easier to to do small PRs and I might get this change that way.

mikelawrence replied:
There have been recently accepted PRs on the sen5x component that change this. You can look at the current dev branch to see these changes. The function has changed to:

``` cpp
// This function performs an in-place conversion of the provided buffer
// from uint16_t values to big endianness
static inline const char *sensirion_convert_to_string_in_place(uint16_t *array, size_t length) {
  for (size_t i = 0; i < length; i++) {
    array[i] = convert_big_endian(array[i]);
  }
  return reinterpret_cast<const char *>(array);
}
```
The calling functions make sure that strings missing a null termination will not cause a problem. Like this:

``` cpp
uint16_t raw_serial_number[8];
const char *serial_number = sensirion_convert_to_string_in_place(raw_serial_number, 8);
snprintf(this->serial_number_, sizeof(this->serial_number_), "%s", serial_number);
```

And this:

```cpp
uint16_t raw_product_name[16];
const char *product_name = sensirion_convert_to_string_in_place(raw_product_name, 16);
ESP_LOGE(TAG, "Product Name does not match: %.32s", product_name);
```

schdro commented on esphome/components/sen5x/sen5x.cpp:
```diff
@@ -43,369 +86,466 @@ static const LogString *rht_accel_mode_to_string(RhtAccelerationMode mode) {
+      if (this->is_sen6x_()) {
+        this->firmware_minor_ = firmware & 0xFF;
+        this->firmware_major_ = firmware >> 8;
+        ESP_LOGV(TAG, "Read Firmware version: %u.%u", this->firmware_major_, this->firmware_minor_);
```
Thought whether such initial info would not warrant as `ESP_LOGD` due to lower frequency.
Because have to enable VERBOSE to see, and then also the repetitive annoying measurements on same level flood trace log later.
From kind of frequency standpoint.

But matter of taste.
Only as observation, no req to change.

mikelawrence replied:
Given that the firmware version is available with the dump I think `ESP_LOGV` is appropriate here.

schdro commented on esphome/components/sen5x/sen5x.h:
```diff
@@ -9,25 +9,32 @@
-  int32_t state0;
-  int32_t state1;
-} PACKED;  // NOLINT
+enum Sen5xType { SEN50, SEN54, SEN55, SEN62, SEN63C, SEN65, SEN66, SEN68, SEN69C, UNKNOWN_MODEL };
```
IMHO type strong `enum class` get more commonly used unless integer compatibility is required.
But certainly also topic of style compat + keeping changeset size limited, as `enum` was also used above.

Only observation, no req to change.

mikelawrence replied:
Not a bad idea but again I was sticking to as few changes as possible. I'll look at this soon

mikelawrence replied:
I'm did a recent PR to improves the existing sen5x component and I added an enum class for the now called `type`

``` cpp
enum class Sen5xType : uint8_t { SEN50, SEN54, SEN55, SEN62, SEN63C, SEN65, SEN66, SEN68, SEN69C, UNKNOWN }
```
However I'm having some trouble getting the config to_code working with an enum class. It tries to set the type to `sen5x::SEN50` instead of the `Sen5xType::SEN50` or `sen5x::Sen5xType::SEN50`. I'm still working this issue though.

mikelawrence replied:
This is now part of the PR.

schdro commented on esphome/components/sen5x/sen5x.h:
```diff
@@ -42,34 +49,44 @@ struct TemperatureCompensation {
+  uint8_t slot;
+
+  TemperatureCompensation() : offset(0), normalized_offset_slope(0), time_constant(0), slot(0) {}
+  TemperatureCompensation(float offset, float normalized_offset_slope, uint16_t time_constant, uint8_t slot = 0) {
```
Guess that (preferred) use of initializer argument is also possible.

Only observation, no req to change.

mikelawrence replied:
I'll look at this soon.

mikelawrence replied:
TemperatureCompensation is an Optional config setting and is defined as an optional class template `optional<TemperatureCompensation> temperature_compensation_`. It appears that optional calls the default initializer.

schdro commented on esphome/components/sen5x/sen5x.h:
```diff
@@ -80,59 +97,74 @@ class SEN5XComponent : public PollingComponent, public sensirion_common::Sensiri
+  void perform_forced_co2_recalibration(uint16_t co2);
 
  protected:
+  bool is_sen6x_();
```
Getter likely could be `const`.

Only observation, no req to change.

mikelawrence replied:
I'm not a professional programmer so sometimes this distinction eludes me. The type (i.e. SEN50 or SEN66) is applied by the config. So until this happens is_sen6x() is technically undefined although the code will produce a value for is_sen6x(). But it can still change. Doesn't that mean this is not a const?

I do believe that by the time the code is actually using this getter it never changes again. Is that good enough to be defined const?

mikelawrence replied:
This now part of the PR

schdro reviewed (approved):
👍 Positive user test result: Having PR running for quite some time with SEN66 FW 4.0, and lately without spotted issues.

(in former 2025, shortly after startup briefly wrong CO2 values around 65536 were indicated during warmup, but that was fixed in previous commit)

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

mikelawrence committed:
Fix incorrect conflict merge

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Adapt to recent and upcoming changes in dev

mikelawrence committed:
Fix clang semantic, remove separator in baseline value

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Another clang-tidy

mikelawrence committed:
Improve line matching
Reorganize update() one less set_timeout()

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence committed:
Keep enum class for Sen5XType

mikelawrence committed:
Make is_sen6x() const

mikelawrence committed:
Merge branch 'dev' into dev

mikelawrence commented:
The devs have decided to go with the new SEN6X component. It's disappointing to do all this work for so long but it is what it is. I like the concept of open source and I liked the thought of contributing to ESPHome but it this the second something like this has happened to me with ESPHome so I doubt I'll try contributing anymore. So long folks!
